etcdserver/apply.go

```diff
@@ -788,3 +803,17 @@ func isGteRange(rangeEnd []byte) bool {
 func noSideEffect(r *pb.InternalRaftRequest) bool {
 	return r.Range != nil || r.AuthUserGet != nil || r.AuthRoleGet != nil
 }
+
+func removeNeedlessRangeReqs(txn *pb.TxnRequest) {
+	f := func(ops []*pb.RequestOp) {
+		for i := 0; i < len(ops); i++ {
```

Comment:
if you want to avoid all the nil checks + keep the alloc-free stuff,
```
j := 0
for i := 0; i < len(ops); i++ {
    if _, ok := ops[i].Request.(*pb.RequestOp_RequestRange); ok {
        continue
    }
    ops[i] = ops[j]
    j++
}
ops = ops[:j]
```